libs/libcommon/include/common/DateLUTImpl.h

```diff
@@ -352,54 +313,205 @@ class DateLUTImpl
+        {
+            if (!week_year_mode && ((first_weekday_mode && weekday != 0) || (!first_weekday_mode && weekday >= 4)))
+                return 0;
+            week_year_mode = 1;
+            (*year)--;
+            first_daynr -= (days = calc_days_in_year(*year));
+            weekday = (weekday + 53 * 7 - days) % 7;
+        }
+
+        if ((first_weekday_mode && weekday != 0) || (!first_weekday_mode && weekday >= 4))
+            days = daynr - (first_daynr + (7 - weekday));
+        else
+            days = daynr - (first_daynr - weekday);
+
+        if (week_year_mode && days >= 52 * 7)
+        {
+            weekday = (weekday + calc_days_in_year(*year)) % 7;
+            if ((!first_weekday_mode && weekday < 4) || (first_weekday_mode && weekday == 0))
+            {
+                (*year)++;
+                return 1;
+            }
+        }
+        return days / 7 + 1;
     }
 
-    inline unsigned toISOWeek(time_t t) const
+    inline unsigned calc_yearWeek(DayNum d, UInt32 week_mode) const
     {
-        return toISOWeek(toDayNum(t));
+        UInt32 year = 0;
+        UInt8 week = calc_week(d, week_mode | WEEK_YEAR, &year);
+        return week + year * 100;
     }
 
-    /// Number of month from some fixed moment in the past (year * 12 + month)
-    inline unsigned toRelativeMonthNum(DayNum d) const
+    /// Calculate week number of WEEK_NEWYEAR_DAY mode
+    /// The week number 1 is the first week in year that contains January 1,
+    inline unsigned calc_newyear_week(DayNum d, bool monday_first_mode, UInt32 * year) const
     {
-        return lut[d].year * 12 + lut[d].month;
+        UInt16 offset_day = monday_first_mode ? 0U : 1U;
+
+        // Checking the week across the year
+        *year = toYear(DayNum(d + 7 - toDayOfWeek(DayNum(d + offset_day))));
+
+        DayNum first_day = makeDayNum(*year, 1, 1);
+        DayNum this_day = d;
+
+        if (monday_first_mode)
+        {
+            // Rounds down a date to the nearest Monday.
+            first_day = toFirstDayNumOfWeek(first_day);
+            this_day = toFirstDayNumOfWeek(d);
+        }
+        else
+        {
+            // Rounds down a date to the nearest Sunday.
+            if (toDayOfWeek(first_day) != 7)
+                first_day = DayNum(first_day - toDayOfWeek(first_day));
+            if (toDayOfWeek(d) != 7)
+                this_day = DayNum(d - toDayOfWeek(d));
+        }
+        return (this_day - first_day) / 7 + 1;
     }
 
-    inline unsigned toRelativeMonthNum(time_t t) const
+    inline unsigned check_week_mode(UInt32 mode) const
     {
-        return toRelativeMonthNum(toDayNum(t));
+        UInt32 week_format = (mode & 7);
+        if (!(week_format & WEEK_MONDAY_FIRST))
+            week_format ^= WEEK_FIRST_WEEKDAY;
+        return week_format;
     }
 
-    inline unsigned toRelativeQuarterNum(DayNum d) const
+    /*
+      Calculate nr of day since year 0 in new date-system (from 1615)
```

Comment:
What's the purpose of this fuinction? We already have `lut` and `years_lut` arrays which allow to do datetime calculations very quickly. And we don't need to support years from 1615 while we are storing dates as an `uint16` number of days since Unix epoch.

Reply:
This code is ported from mysql and I'm trying to change it.